processors/base/writers/trial_and_record.py

```diff
@@ -126,7 +126,7 @@ def _write_record(conn, trial, record_id, source_id=None, source_url=None, is_pr
         'trial_id': trial['id'],
         'source_id': source_id or record['source_id'],
         'source_url': source_url or record['source_url'],
-        'is_primary': is_primary or record.get('is_primary'),
```

Comment:
Nice catch! I think there're 2 more changes we should do:

1. Add a test that triggers this bug, making sure that it's solved (and, more importantly, that it doesn't appear again in the future)
2. The `_write_record()` parameters `source_id`, `source_url` and`is_primary` shouldn't be optional anymore. This method is only used on this file, and we always set those parameters, so it's both useless to have them optional, and confusing.

I think part of this confusion is because we used to support partial updates, where if a field wasn't set in the new data, we would keep what we had in the DB. However, after https://github.com/opentrials/opentrials/issues/643, this behaviour changed. We don't do partial updates anymore. If a field isn't set in the new data, we set it as `null`. We still have a few leftover code of the previous behaviour, though. For example, `'source_id': source_id or record['source_id']` basically means "If the user haven't set a `source_id`, use the current one in the record". This is a partial update. The correct behaviour should be to require the user to define a `source_id` (as we can't set it as `None` because it's required).

So, on my second point up there, the changes would be to require all parameters of `_write_record()` to be set and remove change the pattern `source_id or record['source_id']` to simply `source_id` (same for `source_url`).

Makes sense?